### DISCUSSION TITLE
Custom GUI development question

### DISCUSSION THREAD

**Discussion by CregoryLC** [pybricks/support#1534] [General]
Is pybricks just code interaction with Lego devices or does it support designing a gui to interact with the code? I'd like to use pybricks for code, but would like to develop a custom gui to control how the code interacts with the lego devices.  
FYI... I am 60 years old and haven't wriiten a line of code since the days of visual studio 6.  I hate to think what I am going to have to reteach myself.

**Comment by Debenben** [comment]
I guess it is hard to answer your question without knowing what you are trying to achieve and what hardware you want to use.

In general, pybricks runs on the Lego devices themselves. You can interact with the Lego device running pybricks using bluetooth low energy in the same way you could interact with original Lego devices. Compared to Lego firmware you have more freedom to design such interactions according to your needs and for devices like TechnicHub you can choose to not interact at all.

Finding a suitable bluetooth library (assuming you want to use bluetooth) and a framework to develop the gui is entirely up to you. Do you want to run it on Desktop, Mobile or Embedded device? What kind of operating system? An MS-Office like application or some gameengine? A webserver people can view in their browser?...

**Reply by jaguilar** [reply]
Presumably OP just wants to have something like the PoweredUP! app, but with pybricks coding rather than block-based coding.

**Reply by laurensvalk** [reply]
Thanks both for your answers!

This could be a useful addition in the future. Maybe Pybricks could have a dedicated "tab" to display shapes, widgets, graphs instantiated by the user code. Perhaps some of these could be interactive.

**Reply by Debenben** [reply]
@laurensvalk What about extending jdoodle or jupyter with web bluetooth and usb to connect to pybricks hubs? I do not think that it is a must-have feature, but very useful in my opinion. Compared to that, a dedicated tab with some shapes and widgets sounds like trying to replicate Lego app in more complicated.

**Reply by laurensvalk** [reply]
Notebooks are certainly an interesting possibility and perhaps already doable by using pybricksdev inside the notebook. This is perhaps a bit too advanced for most users though.

Although it's not currently the highest on our list, extending input/output beyond standard input/output text does seem like it could offer a lot of possibilities, with the one in this thread being just one particular application.

Other applications you can think is a view of live sensor information. Or conversely, using a camera for additional input (someone has already done this, for example, but in a separate app).

The common theme here is that I'd ideally like the user code to be in just one place, so that users won't have to learn and implement multiple pieces of code at the same time.

**Comment by CregoryLC** [comment]
I posted my question a bit prematurely as I was not very familiar with Pybricks. After browsing the Pybricks website and documentation, I discovered you offer a block coding interface. That is still not quite what I'm after. My goal is to develop a customizable desktop **program** to control Lego bluetooth devices similar to [The Brick Automation Project (BAP)](https://lgauge.com/article.php?article=trains/gallery/articles/bap), but with Pybricks as the firmware.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Yes, I know... "**app**".  Old school geezer here :-) 

BAP is no longer in development, but had some very interesting concepts. I think by combining those concepts with Pybricks, a very user friendly gui for Pybricks is possible. This would allow the user to control their Lego devices interactively from a computer or presumably whatever device the gui is developed for.  At this point, I need to decide on an ide and then what to use for gui development. After searching for "python gui" it looks like I'll need to familiarize myself with Python, Tkinter, Pybricks and maybe a more current version of Visual Studio. My old laptop recently became unuseable and I'm not sure I want to try to get Visual Studio 6 installed on Windows 11. Plus, it's been years since I've wriitten any code, so it may be time to start fresh with new tools.
_ _
°¿°
\\_/

**Reply by Debenben** [reply]
I am not familiar with the BAP, but after having a short look at the source code it should definitely be possible to create a fork of the BAP project repo and extend it to be able to also connect to hubs running pybricks. As a starting point you could create a copy of Hub.cs class and make it connect to hubs running pybricks by replacing LPF2_HUB characteristic here
https://github.com/tomjcook/BAP/blob/master/LTP.Core/Devices/Hub.cs#L185
with PybricksServiceUUID
https://github.com/pybricks/pybricks-code/blob/master/src/ble-pybricks-service/protocol.ts#L9


Alternatively, if you want to create your own project using python you could take the code from the tutorial https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/ and extend it to include Tkinter library and take it from there.
